README.md
arc-elasticsearch-pipeline-plugin provides the ElasticsearchExtract and ElasticsearchLoad stages.

## Documentation

Full documentation is available here: https://arc.tripl.ai


CHANGELOG.md
# Change Log

## 1.3.0

- update to ElasticSearch 7.6.1
- update to Spark 2.4.5
- update to Arc 2.8.0

## 1.2.0

- add streaming support to `ElasticsearchLoad`
- update to ElasticSearch 7.3.2.

## 1.1.1

- update to Spark 2.4.4
- update to Arc 2.0.1
- update to Scala 2.12.9

## 1.1.0

- change `org.elasticsearch`. `elasticsearch-hadoop` dependency to transitive.

## 1.0.0

- initial release.


src/it/scala/ai/tripl/arc/util/TestUtils.scala
package ai.tripl.arc.util

import java.io.File
import java.sql.Date
import java.sql.Timestamp
import java.time.{ZoneId, ZonedDateTime}
import java.util.ServiceLoader
import scala.collection.JavaConverters._

import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.core._

import org.apache.spark.sql._
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.rdd.RDD
import org.apache.spark.storage.StorageLevel

import ai.tripl.arc.api.API.ARCContext
import ai.tripl.arc.plugins._
import ai.tripl.arc.util.log.LoggerFactory 
import org.apache.log4j.{Level, Logger}

case class KnownData(
    booleanDatum: Boolean, 
    dateDatum: Date, 
    decimalDatum: Decimal, 
    doubleDatum: Double, 
    integerDatum: Integer, 
    longDatum: Long, 
    stringDatum: String, 
    timeDatum: String,
    timestampDatum: Timestamp,
    nullDatum: Null
)

object TestUtils {

    def getLogger()(implicit spark: SparkSession): ai.tripl.arc.util.log.logger.Logger = {
        val loader = ai.tripl.arc.util.Utils.getContextOrSparkClassLoader
        val logger = LoggerFactory.getLogger(spark.sparkContext.applicationId)
        Logger.getLogger("org").setLevel(Level.ERROR)
        Logger.getLogger("breeze").setLevel(Level.ERROR)
        logger
    }

    def getARCContext(isStreaming: Boolean, environment: String = "test", commandLineArguments: Map[String,String] = Map[String,String]()) = {
      val loader = ai.tripl.arc.util.Utils.getContextOrSparkClassLoader

      ARCContext(
        jobId=None, 
        jobName=None, 
        environment=Option(environment), 
        environmentId=None, 
        configUri=None, 
        isStreaming=isStreaming, 
        ignoreEnvironments=false, 
        commandLineArguments=commandLineArguments,
        storageLevel=StorageLevel.MEMORY_AND_DISK_SER,
        immutableViews=false,
        dynamicConfigurationPlugins=ServiceLoader.load(classOf[DynamicConfigurationPlugin], loader).iterator().asScala.toList,
        lifecyclePlugins=ServiceLoader.load(classOf[LifecyclePlugin], loader).iterator().asScala.toList,
        activeLifecyclePlugins=Nil,
        pipelineStagePlugins=ServiceLoader.load(classOf[PipelineStagePlugin], loader).iterator().asScala.toList,
        udfPlugins=ServiceLoader.load(classOf[UDFPlugin], loader).iterator().asScala.toList,
        userData=collection.mutable.Map.empty
      )
    }

    def datasetEquality(expected: DataFrame, actual: DataFrame)(implicit spark: SparkSession): Boolean = {
        import spark.implicits._

        // if both are empty ignore
        if (expected.count != 0 || actual.count != 0) {

            val expectedHashDF = expected.withColumn("_hashLeft", sha2(to_json(struct(expected.columns.sorted.map(col):_*)),512))
            val actualHashDF = actual.withColumn("_hashRight", sha2(to_json(struct(actual.columns.sorted.map(col):_*)),512))

            val transformedDF = expectedHashDF
                .joinWith(actualHashDF, expectedHashDF("_hashLeft") === actualHashDF("_hashRight"), "full")
                .withColumnRenamed("_1", "expected")
                .withColumnRenamed("_2", "actual")

            transformedDF.persist 

            val expectedExceptActual = transformedDF.filter(col("actual").isNull)
            val actualExceptExpected = transformedDF.filter(col("expected").isNull)
            val expectedExceptActualCount = expectedExceptActual.count
            val actualExceptExpectedCount = actualExceptExpected.count     

            if (expectedExceptActualCount != 0 || actualExceptExpectedCount != 0) {
                println("EXPECTED")
                println(expected.schema)
                expected.show(false)
                println("ACTUAL")
                println(actual.schema)
                actual.show(false)

                transformedDF.unpersist
                false
            } else {
                true
            }  
        } else {
            true
        }  
    }

    def getKnownDataset()(implicit spark: SparkSession): DataFrame = {
        import spark.implicits._
        
        val dataset = Seq(
            KnownData(booleanDatum=true, dateDatum=Date.valueOf("2016-12-18"), decimalDatum=Decimal(54.321, 10, 3), doubleDatum=42.4242, integerDatum=17, longDatum=1520828868, stringDatum="test,breakdelimiter", timestampDatum=Timestamp.from(ZonedDateTime.of(2017, 12, 20, 21, 46, 54, 0, ZoneId.of("UTC")).toInstant), timeDatum="12:34:56", nullDatum=null),
            KnownData(booleanDatum=false, dateDatum=Date.valueOf("2016-12-19"), decimalDatum=Decimal(12.345, 10, 3), doubleDatum=21.2121, integerDatum=34, longDatum=1520828123, stringDatum="breakdelimiter,test", timestampDatum=Timestamp.from(ZonedDateTime.of(2017, 12, 29, 17, 21, 49, 0, ZoneId.of("UTC")).toInstant), timeDatum="23:45:16", nullDatum=null)
        )

        dataset.toDF
    }

    // modified dataset for DiffTransform test
    def getKnownAlteredDataset()(implicit spark: SparkSession): DataFrame = {
        import spark.implicits._
        
        val dataset = Seq(
            // same first row
            KnownData(booleanDatum=true, dateDatum=Date.valueOf("2016-12-18"), decimalDatum=Decimal(54.321, 10, 3), doubleDatum=42.4242, integerDatum=17, longDatum=1520828868, stringDatum="test,breakdelimiter", timestampDatum=Timestamp.from(ZonedDateTime.of(2017, 12, 20, 21, 46, 54, 0, ZoneId.of("UTC")).toInstant), timeDatum="12:34:56", nullDatum=null),
            // altered second row (only booleanDatum and integerValue value has been changed)
            KnownData(booleanDatum=true, dateDatum=Date.valueOf("2016-12-19"), decimalDatum=Decimal(12.345, 10, 3), doubleDatum=21.2121, integerDatum=35, longDatum=1520828123, stringDatum="breakdelimiter,test", timestampDatum=Timestamp.from(ZonedDateTime.of(2017, 12, 29, 17, 21, 49, 0, ZoneId.of("UTC")).toInstant), timeDatum="23:45:16", nullDatum=null)
        )

        dataset.toDF
    }    

    def getKnownStringDataset()(implicit spark: SparkSession): DataFrame = {
        val df = getKnownDataset()
        df.select(df.columns.map(c => col(c).cast(StringType)) : _*)
    }    

    def knownDatasetPrettyJSON(row: Int)(implicit spark: SparkSession): String = {
        val json = getKnownDataset().toJSON.collect.toList(row)
        val objectMapper = new ObjectMapper()
        val jsonTree = objectMapper.readTree(json)

        objectMapper.writerWithDefaultPrettyPrinter().writeValueAsString(jsonTree)
    }

    def getListOfFiles(dir: String):List[File] = {
        val d = new File(dir)
        if (d.exists && d.isDirectory) {
            d.listFiles.filter(_.isFile).filter(f => !f.getPath.contains("README.md")).toList
        } else {
            List[File]()
        }
    }     

    def getKnownDatasetMetadataJson(): String = {
    """
    [
        {
            "id": "982cbf60-7ba7-4e50-a09b-d8624a5c49e6",
            "name": "booleanDatum",
            "description": "booleanDatum",
            "type": "boolean",
            "trim": false,
            "nullable": false,
            "nullableValues": [
                "",
                "null"
            ],
            "trueValues": [
                "true"
            ],
            "falseValues": [
                "false"
            ],
            "metadata": {
                "booleanMeta": true,
                "booleanArrayMeta": [true, false],
                "stringMeta": "string",
                "stringArrayMeta": ["string0", "string1"],
                "longMeta": 10,
                "longArrayMeta": [10,20],
                "doubleMeta": 0.141,
                "doubleArrayMeta": [0.141, 0.52],
                "private": false,
                "securityLevel": 0
            }
        },
        {
            "id": "0e8109ba-1000-4b7d-8a4c-b01bae07027f",
            "name": "dateDatum",
            "description": "dateDatum",
            "type": "date",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "formatters": [
                "yyyy-MM-dd"
            ],
            "metadata": {
                "private": true,
                "securityLevel": 3
            }
        },
        {
            "id": "9712c383-22d1-44a6-9ca2-0087af4857f1",
            "name": "decimalDatum",
            "description": "decimalDatum",
            "type": "decimal",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "precision": 38,
            "scale": 18,
            "metadata": {
                "private": true,
                "securityLevel": 2
            }            
        },
        {
            "id": "31541ea3-5b74-4753-857c-770bd601c35b",
            "name": "doubleDatum",
            "description": "doubleDatum",
            "type": "double",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "metadata": {
                "private": true,
                "securityLevel": 8
            }            
        },
        {
            "id": "a66f3bbe-d1c6-44c7-b096-a4be59fdcd78",
            "name": "integerDatum",
            "description": "integerDatum",
            "type": "integer",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "metadata": {
                "private": true,
                "securityLevel": 10
            }            
        },
        {
            "id": "1c0eec1d-17cd-45da-8744-7a9ef5b8b086",
            "name": "longDatum",
            "description": "longDatum",
            "type": "long",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "metadata": {
                "private": false,
                "securityLevel": 0
            }            
        },
        {
            "id": "9712c383-22d1-44a6-9ca2-0087af4857f1",
            "name": "stringDatum",
            "description": "stringDatum",
            "type": "string",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "metadata": {
                "private": false,
                "securityLevel": 0
            }            
        },
        {
            "id": "eb17a18e-4664-4016-8beb-cd2a492d4f20",
            "name": "timeDatum",
            "description": "timeDatum",
            "type": "time",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "formatters": [
                "HH:mm:ss"
            ],
            "metadata": {
                "private": true,
                "securityLevel": 8
            }            
        },        
        {
            "id": "8e42c8f0-22a8-40db-9798-6dd533c1de36",
            "name": "timestampDatum",
            "description": "timestampDatum",
            "type": "timestamp",
            "trim": true,
            "nullable": true,
            "nullableValues": [
                "",
                "null"
            ],
            "formatters": [
                "yyyy-MM-dd'T'HH:mm:ss.SSSXXX"
            ],
            "timezoneId": "UTC",
            "metadata": {
                "private": true,
                "securityLevel": 7
            }            
        }       
    ]
    """
    }
 
}



src/main/scala/ai/tripl/arc/load/ElasticsearchLoad.scala
package ai.tripl.arc.load

import java.net.URI
import java.util.Properties
import scala.collection.JavaConverters._

import org.apache.spark.sql._
import org.apache.spark.sql.types._

import com.typesafe.config._

import org.elasticsearch.spark.sql._ 

import ai.tripl.arc.api._
import ai.tripl.arc.api.API._
import ai.tripl.arc.config._
import ai.tripl.arc.config.Error._
import ai.tripl.arc.plugins.PipelineStagePlugin
import ai.tripl.arc.util.CloudUtils
import ai.tripl.arc.util.DetailException
import ai.tripl.arc.util.EitherUtils._
import ai.tripl.arc.util.ExtractUtils
import ai.tripl.arc.util.MetadataUtils
import ai.tripl.arc.util.ListenerUtils
import ai.tripl.arc.util.Utils
import org.apache.spark.sql.streaming.OutputMode

class ElasticsearchLoad extends PipelineStagePlugin {

  val version = ai.tripl.arc.elasticsearch.BuildInfo.version

  def instantiate(index: Int, config: com.typesafe.config.Config)(implicit spark: SparkSession, logger: ai.tripl.arc.util.log.logger.Logger, arcContext: ARCContext): Either[List[ai.tripl.arc.config.Error.StageError], PipelineStage] = {
    import ai.tripl.arc.config.ConfigReader._
    import ai.tripl.arc.config.ConfigUtils._
    implicit val c = config

    val expectedKeys = "type" :: "name" :: "description" :: "environments"  :: "inputView" :: "output"  :: "numPartitions" :: "partitionBy" :: "saveMode" :: "params" :: "outputMode" :: Nil
    val name = getValue[String]("name")
    val description = getOptionalValue[String]("description")
    val inputView = getValue[String]("inputView")
    val output = getValue[String]("output")
    val numPartitions = getOptionalValue[Int]("numPartitions")
    val partitionBy = getValue[StringList]("partitionBy", default = Some(Nil))
    val saveMode = getValue[String]("saveMode", default = Some("Overwrite"), validValues = "Append" :: "ErrorIfExists" :: "Ignore" :: "Overwrite" :: Nil) |> parseSaveMode("saveMode") _
    val outputMode = getValue[String]("outputMode", default = Some("Append"), validValues = "Append" :: "Complete" :: "Update" :: Nil) |> parseOutputModeType("outputMode") _
    val params = readMap("params", c)
    val invalidKeys = checkValidKeys(c)(expectedKeys)    

    (name, description, inputView, output, numPartitions, partitionBy, saveMode, invalidKeys, outputMode) match {
      case (Right(name), Right(description), Right(inputView), Right(output), Right(numPartitions), Right(partitionBy), Right(saveMode), Right(invalidKeys), Right(outputMode)) => 

        val stage = ElasticsearchLoadStage(
          plugin=this,
          name=name,
          description=description,
          inputView=inputView,
          output=output,
          params=params,
          numPartitions=numPartitions,
          partitionBy=partitionBy,
          saveMode=saveMode,
          outputMode=outputMode
        )

        stage.stageDetail.put("inputView", inputView)  
        stage.stageDetail.put("output", output)  
        stage.stageDetail.put("params", params.asJava)
        stage.stageDetail.put("partitionBy", partitionBy.asJava)
        stage.stageDetail.put("saveMode", saveMode.toString.toLowerCase)
        stage.stageDetail.put("outputMode", outputMode.sparkString)

        Right(stage)
      case _ =>
        val allErrors: Errors = List(name, description, inputView, output, numPartitions, partitionBy, saveMode, invalidKeys, outputMode).collect{ case Left(errs) => errs }.flatten
        val stageName = stringOrDefault(name, "unnamed stage")
        val err = StageError(index, stageName, c.origin.lineNumber, allErrors)
        Left(err :: Nil)
    }
  }
}

case class ElasticsearchLoadStage(
    plugin: ElasticsearchLoad,
    name: String, 
    description: Option[String], 
    inputView: String, 
    output: String, 
    partitionBy: List[String], 
    numPartitions: Option[Int], 
    saveMode: SaveMode, 
    outputMode: OutputModeType,
    params: Map[String, String]
  ) extends PipelineStage {

  override def execute()(implicit spark: SparkSession, logger: ai.tripl.arc.util.log.logger.Logger, arcContext: ARCContext): Option[DataFrame] = {
    ElasticsearchLoadStage.execute(this)
  }
}

object ElasticsearchLoadStage {

  def execute(stage: ElasticsearchLoadStage)(implicit spark: SparkSession, logger: ai.tripl.arc.util.log.logger.Logger, arcContext: ARCContext): Option[DataFrame] = {

    val df = spark.table(stage.inputView)      

    if (!df.isStreaming) {
      stage.numPartitions match {
        case Some(partitions) => stage.stageDetail.put("numPartitions", Integer.valueOf(partitions))
        case None => stage.stageDetail.put("numPartitions", Integer.valueOf(df.rdd.getNumPartitions))
      }
    }

    val dropMap = new java.util.HashMap[String, Object]()

    // elasticsearch cannot support a column called _index
    val unsupported = df.schema.filter(_.name == "_index").map(_.name)
    if (!unsupported.isEmpty) {
      dropMap.put("Unsupported", unsupported.asJava)
    }

    stage.stageDetail.put("drop", dropMap)    
    
    val nonNullDF = df.drop(unsupported:_*)

    val listener = ListenerUtils.addStageCompletedListener(stage.stageDetail)

    // Elasticsearch will convert date and times to epoch milliseconds
    val outputDF = try {
      if (arcContext.isStreaming) {
        nonNullDF.writeStream.options(stage.params).outputMode(stage.outputMode.sparkString).format("es").start(stage.output)
        nonNullDF
      } else {
        stage.partitionBy match {
          case Nil =>
            val dfToWrite = stage.numPartitions.map(nonNullDF.repartition(_)).getOrElse(nonNullDF)
            dfToWrite.write.options(stage.params).mode(stage.saveMode).format("org.elasticsearch.spark.sql").save(stage.output)
            dfToWrite
          case partitionBy => {
            // create a column array for repartitioning
            val partitionCols = partitionBy.map(col => nonNullDF(col))
            stage.numPartitions match {
              case Some(n) =>
                val dfToWrite = nonNullDF.repartition(n, partitionCols:_*)
                dfToWrite.write.options(stage.params).partitionBy(partitionBy:_*).mode(stage.saveMode).format("org.elasticsearch.spark.sql").save(stage.output)
                dfToWrite
              case None =>
                val dfToWrite = nonNullDF.repartition(partitionCols:_*)
                dfToWrite.write.options(stage.params).partitionBy(partitionBy:_*).mode(stage.saveMode).format("org.elasticsearch.spark.sql").save(stage.output)
                dfToWrite
            }
          }
        }
      }
    } catch {
      case e: Exception => throw new Exception(e) with DetailException {
        override val detail = stage.stageDetail
      }
    }

    spark.sparkContext.removeSparkListener(listener)

    Option(outputDF)
  }
}

src/main/scala/ai/tripl/arc/extract/ElasticsearchExtract.scala
package ai.tripl.arc.extract

import java.io._
import java.net.URI
import java.util.Properties

import scala.annotation.tailrec
import scala.collection.JavaConverters._

import org.apache.spark.sql._
import org.apache.spark.sql.functions._
import org.apache.spark.sql.types._
import org.apache.spark.storage.StorageLevel
import org.apache.spark.TaskContext

import org.elasticsearch.spark.sql._ 

import ai.tripl.arc.api._
import ai.tripl.arc.api.API._
import ai.tripl.arc.config._
import ai.tripl.arc.config.Error._
import ai.tripl.arc.plugins.PipelineStagePlugin
import ai.tripl.arc.util.CloudUtils
import ai.tripl.arc.util.DetailException
import ai.tripl.arc.util.EitherUtils._
import ai.tripl.arc.util.ExtractUtils
import ai.tripl.arc.util.MetadataUtils
import ai.tripl.arc.util.Utils

class ElasticsearchExtract extends PipelineStagePlugin {

  val version = ai.tripl.arc.elasticsearch.BuildInfo.version

  def instantiate(index: Int, config: com.typesafe.config.Config)(implicit spark: SparkSession, logger: ai.tripl.arc.util.log.logger.Logger, arcContext: ARCContext): Either[List[ai.tripl.arc.config.Error.StageError], PipelineStage] = {
    import ai.tripl.arc.config.ConfigReader._
    import ai.tripl.arc.config.ConfigUtils._
    implicit val c = config

    val expectedKeys = "type" :: "name" :: "description" :: "environments"  :: "input" :: "outputView"  :: "numPartitions" :: "partitionBy" :: "persist" :: "params" :: Nil
    val name = getValue[String]("name")
    val description = getOptionalValue[String]("description")
    val input = getValue[String]("input")
    val outputView = getValue[String]("outputView")
    val persist = getValue[java.lang.Boolean]("persist", default = Some(false))
    val numPartitions = getOptionalValue[Int]("numPartitions")
    val partitionBy = getValue[StringList]("partitionBy", default = Some(Nil))
    val params = readMap("params", c)
    val invalidKeys = checkValidKeys(c)(expectedKeys)    

    (name, description, input, outputView, persist, numPartitions, partitionBy, invalidKeys) match {
      case (Right(name), Right(description), Right(input), Right(outputView), Right(persist), Right(numPartitions), Right(partitionBy), Right(invalidKeys)) => 

        val stage = ElasticsearchExtractStage(
          plugin=this,
          name=name,
          description=description,
          input=input,
          outputView=outputView,
          params=params,
          persist=persist,
          numPartitions=numPartitions,
          partitionBy=partitionBy
        )

        stage.stageDetail.put("input", input)  
        stage.stageDetail.put("outputView", outputView)  
        stage.stageDetail.put("params", params.asJava)

        Right(stage)
      case _ =>
        val allErrors: Errors = List(name, description, input, outputView, persist, numPartitions, partitionBy, invalidKeys).collect{ case Left(errs) => errs }.flatten
        val stageName = stringOrDefault(name, "unnamed stage")
        val err = StageError(index, stageName, c.origin.lineNumber, allErrors)
        Left(err :: Nil)
    }
  }
}

case class ElasticsearchExtractStage(
    plugin: ElasticsearchExtract,
    name: String, 
    description: Option[String], 
    input: String, 
    outputView: String, 
    params: Map[String, String], 
    persist: Boolean, 
    numPartitions: Option[Int], 
    partitionBy: List[String]
  ) extends PipelineStage {

  override def execute()(implicit spark: SparkSession, logger: ai.tripl.arc.util.log.logger.Logger, arcContext: ARCContext): Option[DataFrame] = {
    ElasticsearchExtractStage.execute(this)
  }
}

object ElasticsearchExtractStage {

  def execute(stage: ElasticsearchExtractStage)(implicit spark: SparkSession, logger: ai.tripl.arc.util.log.logger.Logger, arcContext: ARCContext): Option[DataFrame] = {
    import spark.implicits._

    // if incoming dataset is empty create empty dataset with a known schema
    val df = try {
      if (arcContext.isStreaming) {
        spark.emptyDataFrame
      } else {      
        spark.read.format("org.elasticsearch.spark.sql").options(stage.params).load(stage.input)
      }
    } catch {
      case e: Exception => throw new Exception(e) with DetailException {
        override val detail = stage.stageDetail          
      }
    }    

    // repartition to distribute rows evenly
    val repartitionedDF = stage.partitionBy match {
      case Nil => { 
        stage.numPartitions match {
          case Some(numPartitions) => df.repartition(numPartitions)
          case None => df
        }   
      }
      case partitionBy => {
        // create a column array for repartitioning
        val partitionCols = partitionBy.map(col => df(col))
        stage.numPartitions match {
          case Some(numPartitions) => df.repartition(numPartitions, partitionCols:_*)
          case None => df.repartition(partitionCols:_*)
        }
      }
    } 
    if (arcContext.immutableViews) repartitionedDF.createTempView(stage.outputView) else repartitionedDF.createOrReplaceTempView(stage.outputView)

    if (!repartitionedDF.isStreaming) {
      stage.stageDetail.put("inputFiles", java.lang.Integer.valueOf(repartitionedDF.inputFiles.length))
      stage.stageDetail.put("outputColumns", java.lang.Integer.valueOf(repartitionedDF.schema.length))
      stage.stageDetail.put("numPartitions", java.lang.Integer.valueOf(repartitionedDF.rdd.partitions.length))

      if (stage.persist) {
        repartitionedDF.persist(arcContext.storageLevel)
        stage.stageDetail.put("records", java.lang.Long.valueOf(repartitionedDF.count)) 
      }      
    }

    Option(repartitionedDF)
  }

}



src/it/scala/ai/tripl/arc/extract/ElasticsearchExtractSuite.scala
package ai.tripl.arc

import java.net.URI

import org.scalatest.FunSuite
import org.scalatest.BeforeAndAfter

import collection.JavaConverters._

import org.apache.http.client.methods.HttpDelete
import org.apache.http.impl.client.HttpClientBuilder

import org.apache.commons.io.FileUtils
import org.apache.commons.io.IOUtils
import org.apache.spark.sql._
import org.apache.spark.sql.functions._

import ai.tripl.arc.api._
import ai.tripl.arc.api.API._
import ai.tripl.arc.util._
import ai.tripl.arc.util.ControlUtils._

import org.elasticsearch.spark.sql._ 

class ElasticsearchExtractSuite extends FunSuite with BeforeAndAfter {

  var session: SparkSession = _  

  val testData = getClass.getResource("/akc_breed_info.csv").toString
  val outputView = "actual"
  val index = "dogs"
  val esURL = "elasticsearch"
  val port = "9200"
  val wanOnly = "true"
  val ssl = "false"  

  before {
    implicit val spark = SparkSession
                  .builder()
                  .master("local[*]")
                  .config("spark.ui.port", "9999")
                  .appName("Spark ETL Test")
                  .getOrCreate()
    spark.sparkContext.setLogLevel("INFO")
    implicit val logger = TestUtils.getLogger()

    // set for deterministic timezone
    spark.conf.set("spark.sql.session.timeZone", "UTC")   

    session = spark
  }

  after {
    session.stop
  }

  test("ElasticsearchExtract") {
    implicit val spark = session
    import spark.implicits._
    implicit val logger = TestUtils.getLogger()
    implicit val arcContext = TestUtils.getARCContext(isStreaming=false)

    val df0 = spark.read.option("header","true").csv(testData)
    df0.createOrReplaceTempView("expected")

    val client = HttpClientBuilder.create.build
    val delete = new HttpDelete(s"http://${esURL}:9200/index")
    val response = client.execute(delete)
    response.close 

    df0.write
      .format("org.elasticsearch.spark.sql")
      .option("es.nodes.wan.only",wanOnly)
      .option("es.port",port)
      .option("es.net.ssl",ssl)
      .option("es.nodes", esURL)
      .mode("overwrite")
      .save(index)

    extract.ElasticsearchExtractStage.execute(
      extract.ElasticsearchExtractStage(
        plugin=new extract.ElasticsearchExtract,
        name="df", 
        description=None,
        input=index,
        outputView=outputView, 
        numPartitions=None,
        params=Map("es.nodes.wan.only" -> wanOnly, "es.port" -> port, "es.net.ssl" -> ssl, "es.nodes" -> esURL),
        partitionBy=Nil,
        persist=true
      )
    )   

    // reselect fields to ensure correct order
    val expected = spark.sql(s"""
    SELECT Breed, height_high_inches, height_low_inches, weight_high_lbs, weight_low_lbs FROM expected
    """)

    // reselect fields to ensure correct order
    val actual = spark.sql(s"""
    SELECT Breed, height_high_inches, height_low_inches, weight_high_lbs, weight_low_lbs FROM ${outputView}
    """)

    val actualExceptExpectedCount = actual.except(expected).count
    val expectedExceptActualCount = expected.except(actual).count
    if (actualExceptExpectedCount != 0 || expectedExceptActualCount != 0) {
      println("actual")
      actual.show(100000, false)
      println("expected")
      expected.show(100000, false)  
    }
    assert(actualExceptExpectedCount === 0)
    assert(expectedExceptActualCount === 0)
  }    

}


src/it/scala/ai/tripl/arc/load/ElasticsearchLoadSuite.scala
package ai.tripl.arc

import java.net.URI
import java.util.UUID
import java.util.Properties
import scala.util.Random
import org.apache.http.client.methods.{HttpDelete, HttpGet}
import org.apache.http.impl.client.HttpClientBuilder
import com.fasterxml.jackson.databind.ObjectMapper

import org.scalatest.FunSuite
import org.scalatest.BeforeAndAfter

import scala.collection.JavaConverters._
import scala.io.Source

import org.apache.commons.io.FileUtils
import org.apache.commons.io.IOUtils
import org.apache.spark.sql._
import org.apache.spark.sql.functions._

import ai.tripl.arc.api._
import ai.tripl.arc.api.API._
import ai.tripl.arc.util._
import ai.tripl.arc.config.ArcPipeline
import org.elasticsearch.spark.sql._

class ElasticsearchLoadSuite extends FunSuite with BeforeAndAfter {


  val alpha = "abcdefghijklmnopqrstuvwxyz"
  val size = alpha.size
  def randStr(n:Int) = (1 to n).map(x => alpha(Random.nextInt.abs % size)).mkString

  var session: SparkSession = _
  val testData = getClass.getResource("/akc_breed_info.csv").toString
  val inputView = "expected"
  val index = "dogs"
  val esURL = "elasticsearch"
  val port = "9200"
  val wanOnly = "true"
  val ssl = "false"
  val streamingIndex = "streaming"
  val outputView = "outputView"
  val checkpointLocation = "/tmp/checkpointLocation"

  before {
    implicit val spark = SparkSession
                  .builder()
                  .master("local[*]")
                  .config("spark.ui.port", "9999")
                  .config("es.index.auto.create", "true")
                  .config("spark.sql.streaming.checkpointLocation", checkpointLocation)
                  .appName("Spark ETL Test")
                  .getOrCreate()
    spark.sparkContext.setLogLevel("INFO")
    implicit val logger = TestUtils.getLogger()

    // set for deterministic timezone
    spark.conf.set("spark.sql.session.timeZone", "UTC")

    session = spark
  }

  after {
    session.stop
  }

  test("ElasticsearchLoad") {
    implicit val spark = session
    import spark.implicits._
    implicit val logger = TestUtils.getLogger()
    implicit val arcContext = TestUtils.getARCContext(isStreaming=false)

    val df0 = spark.read.option("header","true").csv(testData)
    df0.createOrReplaceTempView(inputView)

    val client = HttpClientBuilder.create.build
    val delete = new HttpDelete(s"http://${esURL}:9200/${index}")
    val response = client.execute(delete)
    response.close

    load.ElasticsearchLoadStage.execute(
      load.ElasticsearchLoadStage(
        plugin=new load.ElasticsearchLoad,
        name="df",
        description=None,
        inputView=inputView,
        output=index,
        numPartitions=None,
        params=Map("es.nodes.wan.only" -> wanOnly, "es.port" -> port, "es.net.ssl" -> ssl, "es.nodes" -> esURL),
        saveMode=SaveMode.Overwrite,
        outputMode=OutputModeTypeAppend,
        partitionBy=Nil
      )
    )

    val df1 = spark.read
      .format("org.elasticsearch.spark.sql")
      .option("es.nodes.wan.only",wanOnly)
      .option("es.port", port)
      .option("es.net.ssl", ssl)
      .option("es.nodes", esURL)
      .load(index)

    df1.createOrReplaceTempView("actual")


    // reselect fields to ensure correct order
    val expected = spark.sql(s"""
    SELECT Breed, height_high_inches, height_low_inches, weight_high_lbs, weight_low_lbs FROM ${inputView}
    """)

    // reselect fields to ensure correct order
    val actual = spark.sql(s"""
    SELECT Breed, height_high_inches, height_low_inches, weight_high_lbs, weight_low_lbs FROM actual
    """)

    val actualExceptExpectedCount = actual.except(expected).count
    val expectedExceptActualCount = expected.except(actual).count
    if (actualExceptExpectedCount != 0 || expectedExceptActualCount != 0) {
      println("actual")
      actual.show(100000, false)
      println("expected")
      expected.show(100000, false)
    }
    assert(actualExceptExpectedCount === 0)
    assert(expectedExceptActualCount === 0)
  }

  test("ElasticsearchLoad end-to-end") {
    implicit val spark = session
    import spark.implicits._
    implicit val logger = TestUtils.getLogger()
    implicit val arcContext = TestUtils.getARCContext(isStreaming=false)

    val df = spark.read.option("header","true").csv(testData)
    df.createOrReplaceTempView(inputView)

    val conf = s"""{
      "stages": [
        {
          "type": "ElasticsearchLoad",
          "name": "write person",
          "environments": [
            "production",
            "test"
          ],
          "output": "person",
          "inputView": "${inputView}",
          "saveMode": "Overwrite",
          "params": {
            "es.nodes": "${esURL}",
            "es.port": "${port}",
            "es.nodes.wan.only": "${wanOnly}",
            "es.net.ssl": "${ssl}"
          }
        }
      ]
    }"""

    val pipelineEither = ArcPipeline.parseConfig(Left(conf), arcContext)

    pipelineEither match {
      case Left(_) => {
        println(pipelineEither)
        assert(false)
      }
      case Right((pipeline, _)) => ARC.run(pipeline)(spark, logger, arcContext)
    }
  }

  test("ElasticsearchLoad: Structured Streaming") {
    implicit val spark = session
    import spark.implicits._
    implicit val logger = TestUtils.getLogger()
    implicit val arcContext = TestUtils.getARCContext(isStreaming=true)
    FileUtils.deleteQuietly(new java.io.File(checkpointLocation))

    val indexName = randStr(10)

    val readStream = spark
      .readStream
      .format("rate")
      .option("rowsPerSecond", "1")
      .load

    readStream.createOrReplaceTempView(inputView)

    load.ElasticsearchLoadStage.execute(
      load.ElasticsearchLoadStage(
        plugin=new load.ElasticsearchLoad,
        name="df",
        description=None,
        inputView=inputView,
        output=indexName,
        numPartitions=None,
        params=Map("es.nodes.wan.only" -> wanOnly, "es.port" -> port, "es.net.ssl" -> ssl, "es.nodes" -> esURL),
        saveMode=SaveMode.Overwrite,
        outputMode=OutputModeTypeAppend,
        partitionBy=Nil
      )
    )
    Thread.sleep(2000)

    spark.streams.active.foreach(streamingQuery => streamingQuery.stop)

    // call _search rest api to get all documents for new index
    val client = HttpClientBuilder.create.build
    val get = new HttpGet(s"http://${esURL}:${port}/${indexName}/_search")
    val response = client.execute(get)
    val body = Source.fromInputStream(response.getEntity.getContent).mkString
    response.close

    // assert that the documents array returned in the search is not empty
    // if no documents then hits.hits will fail anyway
    spark.read.json(spark.sparkContext.parallelize(Seq(body)).toDF.as[String]).createOrReplaceTempView("response")
    val hitsSize = spark.sql("""
    SELECT SIZE(hits.hits) FROM response
    """)
    assert(hitsSize.first.getInt(0) != 0)
  }
}